subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt

```diff
@@ -0,0 +1,236 @@
+                }
+            """
+        )
+        configureProject().buildScriptCompiled().andOutputContains("foo")
+
+        withPrecompiledScriptPluginInBuildSrc(
+            pluginId,
+            """
+                println("bar")
+            """
+        )
+        configureProject().buildScriptNotCompiled().andOutputContains("bar")
+    }
+
+    private
+    fun withPrecompiledScriptPluginInBuildSrc(pluginId: String, pluginSource: String) {
+        withKotlinDslPluginIn("buildSrc")
+        withFile("buildSrc/src/main/kotlin/$pluginId.gradle.kts", pluginSource)
+    }
+
+    private
+    fun buildJarForBuildScriptClasspath(classBody: String): Pair<String, String> {
+        val baseDir = "buildscript"
+        withSettingsIn(
+            baseDir,
+            """
+                rootProject.name = "buildscript"
+            """
+        )
+        withBuildScriptIn(
+            baseDir,
+            """
+                plugins {
+                    id("java-library")
+                }
+            """
+        )
+        val className = sourceFile(baseDir, classBody)
+        build(existing(baseDir), "build")
+        val jarPath = "$baseDir/build/libs/buildscript.jar"
+        assertTrue(existing(jarPath).exists())
+        return Pair(className, jarPath)
+    }
+
+    private
+    fun givenClassInBuildSrcContains(classBody: String): String =
+        sourceFile("buildSrc", classBody)
+
+    private
+    fun sourceFile(baseDir: String, classBody: String): String {
+        withFile(
+            "$baseDir/src/main/java/com/example/Foo.java",
+            """
+            package com.example;
+            public class Foo {
+                $classBody
+            }
+            """
+        )
+        return "com.example.Foo"
+    }
+
+    private
+    fun configureProject() =
+        OutputFixture(gradleExecuterFor(arrayOf("--debug")).withStackTraceChecksDisabled().run().output)
+
+    class OutputFixture(val output: String) {
+        private
+        val scriptClasspathCompileOperationStartMarker = "Build operation 'Compile script build.gradle.kts (CLASSPATH)' started"
+
+        private
+        val scriptBodyCompileOperationStartMarker = "Build operation 'Compile script build.gradle.kts (BODY)' started"
+
+        fun buildScriptCompiled(): OutputFixture {
+            if (!output.contains(scriptClasspathCompileOperationStartMarker) && !output.contains(scriptBodyCompileOperationStartMarker)) {
+                throw AssertionError("Expected script to be compiled, but it wasn't")
+            }
+            return this
+        }
+
+        fun buildScriptNotCompiled(): OutputFixture {
+            if (output.contains(scriptClasspathCompileOperationStartMarker) || output.contains(scriptBodyCompileOperationStartMarker)) {
+                throw AssertionError("Expected script compilation to be avoided, but the buildscript was recompiled")
+            }
+            return this
+        }
```

Comment:
There has to be a better way to verify whether the build scripts were compiled or reused from cache than relying on `--debug` output.